From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin Raneri <kevin.raneri@gmail.com>
Date: Wed, 27 Dec 2023 22:49:53 -0500
Subject: [PATCH] TPS catchup


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index be53079a9965a79c459ae7a769b375033406bd58..3c6962eabeb9cbbbdd9d3a814e890a3d8da102b8 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1262,6 +1262,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                     profilerFiller.popPush("nextTickWait");
                     this.mayHaveDelayedTasks = true;
                     this.delayedTasksMaxNextTickTimeNanos = Math.max(Util.getNanos() + l, this.nextTickTimeNanos);
+                    // Pufferfish start - tps catchup
+                    if (!gg.pufferfish.pufferfish.PufferfishConfig.tpsCatchup) {
+                        this.nextTickTimeNanos = currentTime + l;
+                        this.delayedTasksMaxNextTickTimeNanos = nextTickTimeNanos;
+                    }
+                    // Pufferfish end
                     this.startMeasuringTaskExecutionTime();
                     this.waitUntilNextTick();
                     this.finishMeasuringTaskExecutionTime();
