--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -2728,7 +_,11 @@
     @Override
     public void swing(final InteractionHand hand) {
         super.swing(hand);
-        this.resetAttackStrengthTicker();
+        // DivineMC start - Fix MC-118740
+        if (hand == InteractionHand.MAIN_HAND && org.bxteam.divinemc.config.DivineConfig.FixesCategory.fixMc118740) {
+            this.resetAttackStrengthTicker();
+        }
+        // DivineMC end - Fix MC-118740
     }
 
     public boolean isChangingDimension() {
