From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sat, 26 Apr 2025 23:43:59 +0300
Subject: [PATCH] Optimize default values for configs


diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index 025ec3f3e84a3367b4d709c8c3ecd98f6dfecab4..2d980416885ec4b39aaf74509b5519f86ab29195 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -284,7 +284,7 @@ public class SpigotConfig {
 
     public static boolean saveUserCacheOnStopOnly;
     private static void saveUserCacheOnStopOnly() {
-        SpigotConfig.saveUserCacheOnStopOnly = SpigotConfig.getBoolean("settings.save-user-cache-on-stop-only", false);
+        SpigotConfig.saveUserCacheOnStopOnly = SpigotConfig.getBoolean("settings.save-user-cache-on-stop-only", true); // DivineMC - Optimize default values for configs - usercache is a rebuildable cache, no need to rewrite it on every disconnect
     }
 
     public static double movedWronglyThreshold;
diff --git a/src/main/resources/configurations/bukkit.yml b/src/main/resources/configurations/bukkit.yml
index 00a5b4dc0d7bc57ae3f9231f84a081617ec4b15b..bdaa083be991aafe12e62cda71c3c5839e387a7e 100644
--- a/src/main/resources/configurations/bukkit.yml
+++ b/src/main/resources/configurations/bukkit.yml
@@ -16,7 +16,7 @@ settings:
     update-folder: update
     plugin-profiling: false
     connection-throttle: 4000
-    query-plugins: true
+    query-plugins: false
     deprecated-verbose: default
     shutdown-message: Server closed
     minimum-api: none
diff --git a/src/main/resources/configurations/commands.yml b/src/main/resources/configurations/commands.yml
index ddf2cd4b802cf7d22342eed7f09c989761aed7a9..59f2bed6512f23ca7e28aebe0cef7d50bcfb8f8c 100644
--- a/src/main/resources/configurations/commands.yml
+++ b/src/main/resources/configurations/commands.yml
@@ -10,6 +10,4 @@
 
 command-block-overrides: []
 ignore-vanilla-permissions: false
-aliases:
-    icanhasbukkit:
-      - "version $1-"
+aliases: []
