From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sun, 23 Mar 2025 16:47:24 +0300
Subject: [PATCH] Paper PR: Add FillBottleEvents for player and dispenser


diff --git a/net/minecraft/core/cauldron/CauldronInteractions.java b/net/minecraft/core/cauldron/CauldronInteractions.java
index 04b47a788586d21a37f73ba2dcf9036ea174872b..fbc3a7ef33a6db96e97dbfd146e5e0d29e92b324 100644
--- a/net/minecraft/core/cauldron/CauldronInteractions.java
+++ b/net/minecraft/core/cauldron/CauldronInteractions.java
@@ -56,7 +56,12 @@ public class CauldronInteractions {
                     }
                     // CraftBukkit end
                     Item usedItem = itemInHand.getItem();
-                    player.setItemInHand(hand, ItemUtils.createFilledResult(itemInHand, player, new ItemStack(Items.GLASS_BOTTLE)));
+                    // DivineMC start - Paper PR: Add FillBottleEvents for player and dispenser
+                    final io.papermc.paper.event.player.PlayerFillBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFillBottleEvent(player, hand, itemInHand, ItemUtils.createFilledResult(itemInHand, player, new ItemStack(Items.GLASS_BOTTLE)));
+                    if (event.isCancelled()) {
+                        return InteractionResult.PASS;
+                    }
+                    // DivineMC end - Paper PR: Add FillBottleEvents for player and dispenser
                     player.awardStat(Stats.USE_CAULDRON);
                     player.awardStat(Stats.ITEM_USED.get(usedItem));
                     // level.setBlockAndUpdate(pos, Blocks.WATER_CAULDRON.defaultBlockState()); // CraftBukkit
diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 56b9f9016a652a704d785de683738efcf02a44f9..d00caca4e8b593e821c2f19de091d000180f5346 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -523,13 +523,25 @@ public interface DispenseItemBehavior {
                     // Paper end - Call BlockDispenseEvent
                     if (state.is(BlockTags.BEEHIVES, s -> s.hasProperty(BeehiveBlock.HONEY_LEVEL) && s.getBlock() instanceof BeehiveBlock)
                         && state.getValue(BeehiveBlock.HONEY_LEVEL) >= 5) {
+                        // DivineMC start - Paper PR: Add FillBottleEvents for player and dispenser
+                        final io.papermc.paper.event.block.BlockFillBottleEvent bottleEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFillBottleEvent(level, source.pos(), dispensed, new ItemStack(Items.HONEY_BOTTLE));
+                        if (bottleEvent.isCancelled()) {
+                            return dispensed;
+                        }
+                        // DivineMC end - Paper PR: Add FillBottleEvents for player and dispenser
                         ((BeehiveBlock)state.getBlock())
                             .releaseBeesAndResetHoneyLevel(level, state, target, null, BeehiveBlockEntity.BeeReleaseStatus.BEE_RELEASED);
                         this.setSuccess(true);
-                        return this.takeLiquid(source, dispensed, new ItemStack(Items.HONEY_BOTTLE));
+                        return this.takeLiquid(source, dispensed, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(bottleEvent.getResultItem())); // DivineMC - Paper PR: Add FillBottleEvents for player and dispenser
                     } else if (level.getFluidState(target).is(FluidTags.WATER)) {
+                        // DivineMC start - Paper PR: Add FillBottleEvents for player and dispenser
+                        final io.papermc.paper.event.block.BlockFillBottleEvent bottleEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFillBottleEvent(level, source.pos(), dispensed, PotionContents.createItemStack(Items.POTION, Potions.WATER));
+                        if (bottleEvent.isCancelled()) {
+                            return dispensed;
+                        }
+                        // DivineMC end - Paper PR: Add FillBottleEvents for player and dispenser
                         this.setSuccess(true);
-                        return this.takeLiquid(source, dispensed, PotionContents.createItemStack(Items.POTION, Potions.WATER));
+                        return this.takeLiquid(source, dispensed, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(bottleEvent.getResultItem())); // DivineMC - Paper PR: Add FillBottleEvents for player and dispenser
                     } else {
                         return super.execute(source, dispensed);
                     }
diff --git a/net/minecraft/world/item/BottleItem.java b/net/minecraft/world/item/BottleItem.java
index e45fb987b709cbd798b7bcd179a559d8ab70e6ff..a0a84f7b97d94989768beee39206b807e21a0f26 100644
--- a/net/minecraft/world/item/BottleItem.java
+++ b/net/minecraft/world/item/BottleItem.java
@@ -33,6 +33,18 @@ public class BottleItem extends Item {
         );
         ItemStack itemStack = player.getItemInHand(hand);
         if (!clouds.isEmpty()) {
+            // DivineMC start - Paper PR: Add FillBottleEvents for player and dispenser
+            final io.papermc.paper.event.player.PlayerFillBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFillBottleEvent(player, hand, itemStack, new ItemStack(Items.DRAGON_BREATH));
+            //noinspection DuplicatedCode
+            if (event.isCancelled()) {
+                player.containerMenu.sendAllDataToRemote();
+                return InteractionResult.PASS;
+            }
+            final ItemStack resultItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem());
+            if (resultItem.is(itemStack.getItem())) {
+                player.containerMenu.sendAllDataToRemote();
+            }
+            // DivineMC end - Paper PR: Add FillBottleEvents for player and dispenser
             AreaEffectCloud cloud = clouds.get(0);
             cloud.setRadius(cloud.getRadius() - 0.5F);
             level.playSound(null, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL_DRAGONBREATH, SoundSource.NEUTRAL, 1.0F, 1.0F);
@@ -41,7 +53,7 @@ public class BottleItem extends Item {
                 CriteriaTriggers.PLAYER_INTERACTED_WITH_ENTITY.trigger(serverPlayer, itemStack, cloud);
             }
 
-            return InteractionResult.SUCCESS.heldItemTransformedTo(this.turnBottleIntoItem(itemStack, player, new ItemStack(Items.DRAGON_BREATH)));
+            return InteractionResult.SUCCESS.heldItemTransformedTo(this.turnBottleIntoItem(itemStack, player, resultItem)); // DivineMC - Paper PR: Add FillBottleEvents for player and dispenser
         } else {
             BlockHitResult hitResult = getPlayerPOVHitResult(level, player, ClipContext.Fluid.SOURCE_ONLY);
             if (hitResult.getType() == HitResult.Type.MISS) {
@@ -55,10 +67,22 @@ public class BottleItem extends Item {
                 }
 
                 if (level.getFluidState(pos).is(FluidTags.WATER)) {
+                        // DivineMC start - Paper PR: Add FillBottleEvents for player and dispenser
+                        final io.papermc.paper.event.player.PlayerFillBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFillBottleEvent(player, hand, itemStack, PotionContents.createItemStack(Items.POTION, Potions.WATER));
+                        //noinspection DuplicatedCode
+                        if (event.isCancelled()) {
+                            player.containerMenu.sendAllDataToRemote();
+                            return InteractionResult.PASS;
+                        }
+                        final ItemStack resultItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem());
+                        if (resultItem.is(itemStack.getItem())) {
+                            player.containerMenu.sendAllDataToRemote();
+                        }
+                        // DivineMC end - Paper PR: Add FillBottleEvents for player and dispenser
                     level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL, SoundSource.NEUTRAL, 1.0F, 1.0F);
                     level.gameEvent(player, GameEvent.FLUID_PICKUP, pos);
                     return InteractionResult.SUCCESS
-                        .heldItemTransformedTo(this.turnBottleIntoItem(itemStack, player, PotionContents.createItemStack(Items.POTION, Potions.WATER)));
+                            .heldItemTransformedTo(this.turnBottleIntoItem(itemStack, player, resultItem)); // DivineMC - Paper PR: Add FillBottleEvents for player and dispenser
                 }
             }
 
diff --git a/net/minecraft/world/level/block/BeehiveBlock.java b/net/minecraft/world/level/block/BeehiveBlock.java
index 0d3e97012b0d590dc2b93e4886a7009f6f78e0d7..360875c9b6f59d091d587e7f93255bff0dd5f7ec 100644
--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -180,12 +180,26 @@ public class BeehiveBlock extends BaseEntityBlock {
                 hiveEmptied = true;
                 level.gameEvent(player, GameEvent.SHEAR, pos);
             } else if (itemStack.is(Items.GLASS_BOTTLE)) {
+                // DivineMC start - Paper PR: Add FillBottleEvents for player and dispenser
+                final io.papermc.paper.event.player.PlayerFillBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFillBottleEvent(player, hand, itemStack, new ItemStack(Items.HONEY_BOTTLE));
+                //noinspection DuplicatedCode
+                if (event.isCancelled()) {
+                    player.containerMenu.sendAllDataToRemote();
+                    return InteractionResult.PASS;
+                }
+                final ItemStack resultItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem());
+                if (resultItem.is(itemStack.getItem())) {
+                    player.containerMenu.sendAllDataToRemote();
+                }
+                // DivineMC end - Paper PR: Add FillBottleEvents for player and dispenser
                 itemStack.shrink(1);
                 level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL, SoundSource.BLOCKS, 1.0F, 1.0F);
                 if (itemStack.isEmpty()) {
-                    player.setItemInHand(hand, new ItemStack(Items.HONEY_BOTTLE));
-                } else if (!player.getInventory().add(new ItemStack(Items.HONEY_BOTTLE))) {
-                    player.drop(new ItemStack(Items.HONEY_BOTTLE), false);
+                    // DivineMC start - Paper PR: Add FillBottleEvents for player and dispenser
+                    player.setItemInHand(hand, resultItem);
+                } else if (!player.getInventory().add(resultItem)) {
+                    player.drop(resultItem, false);
+                    // DivineMC end - Paper PR: Add FillBottleEvents for player and dispenser
                 }
 
                 hiveEmptied = true;
