From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 2 Aug 2025 20:35:47 -0500
Subject: [PATCH] Leaf Fix Enderpearl Teleport


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index f1a8166b09b985aba4e8a501d3c87f5e3eb9e00f..129818dd1b51ebbc2755188450809eec1464e420 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -108,6 +108,33 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         // Store pre teleportation position as the teleport has been moved up.
                         final double preTeleportX = player.getX(), preTeleportY = player.getY(), preTeleportZ = player.getZ();
                         final float preTeleportYRot = player.getYRot(), preTeleportXRot = player.getXRot();
+                        // Leaf start - Fix teleport cross-dimension
+                        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) {
+                            player.getBukkitEntity().taskScheduler.schedule((ServerPlayer taskPlayer) -> {
+                                ServerPlayer newOwner = player.teleport(new TeleportTransition(
+                                    level, teleportPos, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
+                                if (newOwner == null) {
+                                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
+                                    return;
+                                }
+                                if (this.random.nextFloat() < 0.05F && level.isSpawningMonsters()) {
+                                    Endermite endermite = EntityTypes.ENDERMITE.create(level, EntitySpawnReason.TRIGGERED);
+                                    if (endermite != null) {
+                                        endermite.snapTo(preTeleportX, preTeleportY, preTeleportZ, preTeleportYRot, preTeleportXRot); // Paper - spawn endermite at pre teleport position as teleport has been moved up
+                                        level.addFreshEntity(endermite, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.ENDER_PEARL); // Paper - add reason
+                                    }
+                                }
+                                if (this.isOnPortalCooldown()) {
+                                    owner.setPortalCooldown();
+                                }
+                                if (newOwner != null) {
+                                    newOwner.resetFallDistance();
+                                    newOwner.resetCurrentImpulseContext();
+                                    newOwner.hurtServer(player.level(), this.damageSources().enderPearl().eventEntityDamager(this), 5.0F); // CraftBukkit // Paper - fix DamageSource API
+                                }
+                                this.playSound(level, teleportPos);
+                            }, e -> {}, 0);
+                        } else { // Leaf end - Fix teleport cross-dimension
                         ServerPlayer newOwner = player.teleport(new TeleportTransition(
                             level, teleportPos, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
                         if (newOwner == null) {
@@ -141,6 +168,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                         }
 
                         this.playSound(level, teleportPos);
+                        } // Leaf - Fix teleport cross-dimension
                     }
                 } else {
                     Entity newOwner = owner.teleport(
