From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sun, 3 Aug 2025 22:59:22 -0500
Subject: [PATCH] Leaf Fix Too Many Chained Neighbors


diff --git a/io/papermc/paper/redstone/RedstoneWireTurbo.java b/io/papermc/paper/redstone/RedstoneWireTurbo.java
index 8e03419df1469c6fecfaa82fcb319a58635b39f3..b40935d10a6cdb733fd99dc7fd6a3940a0e9a21d 100644
--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -827,14 +827,16 @@ public final class RedstoneWireTurbo {
         j = getMaxCurrentStrength(upd, j);
         int l = 0;
 
+        final int k; if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) k = wire.getBlockSignal(worldIn, upd.self); else { // Leaf - Fix too many chained neighbors
         wire.shouldSignal = false;
         // Unfortunately, World.isBlockIndirectlyGettingPowered is complicated,
         // and I'm not ready to try to replicate even more functionality from
         // elsewhere in Minecraft into this accelerator.  So sadly, we must
         // suffer the performance hit of this very expensive call.  If there
         // is consistency to what this call returns, we may be able to cache it.
-        final int k = worldIn.getBestNeighborSignal(upd.self);
+        k = worldIn.getBestNeighborSignal(upd.self);
         wire.shouldSignal = true;
+        } // Leaf - Fix too many chained neighbors
 
         // The variable 'k' holds the maximum redstone power value of any adjacent blocks.
         // If 'k' has the highest level of all neighbors, then the power level of this
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index 3fb9600b0301503e9593915a9d96ddbe2863e2f1..eaf9ad7c22e3473386b579de7eaa572775d47409 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -66,6 +66,7 @@ public class RedStoneWireBlock extends Block {
     private final BlockState crossState;
     private final RedstoneWireEvaluator evaluator = new DefaultRedstoneWireEvaluator(this);
     public boolean shouldSignal = true;
+    private final ThreadLocal<Boolean> fish$shouldSignalTL = ThreadLocal.withInitial(() -> true);
 
     @Override
     public MapCodec<RedStoneWireBlock> codec() {
@@ -336,6 +337,12 @@ public class RedStoneWireBlock extends Block {
     }
 
     public int getBlockSignal(final Level level, final BlockPos pos) {
+        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) {
+            this.fish$shouldSignalTL.set(false);
+            int bestNeighborSignal = level.getBestNeighborSignal(pos);
+            this.fish$shouldSignalTL.set(true);
+            return bestNeighborSignal;
+        }
         this.shouldSignal = false;
         int blockSignal = level.getBestNeighborSignal(pos);
         this.shouldSignal = true;
@@ -432,12 +439,14 @@ public class RedStoneWireBlock extends Block {
 
     @Override
     protected int getDirectSignal(final BlockState state, final BlockGetter level, final BlockPos pos, final Direction direction) {
+        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) return !this.fish$shouldSignalTL.get() ? 0 : state.getSignal(level, pos, direction);
         return !this.shouldSignal ? Redstone.SIGNAL_MIN : state.getSignal(level, pos, direction);
     }
 
     @Override
     protected int getSignal(final BlockState state, final BlockGetter level, final BlockPos pos, final Direction direction) {
-        if (this.shouldSignal && direction != Direction.DOWN) {
+        boolean shouldSignal0 = me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED ? this.fish$shouldSignalTL.get() : this.shouldSignal;
+        if (shouldSignal0 && direction != Direction.DOWN) {
             int power = this.ownSignal(state, level, pos);
             if (power == 0) {
                 return Redstone.SIGNAL_MIN;
@@ -474,6 +483,7 @@ public class RedStoneWireBlock extends Block {
 
     @Override
     protected boolean isSignalSource(final BlockState state) {
+        if (me.biquaternions.fish.FishConfig.ASYNC.WORLD_TICKING._ENABLED) return this.fish$shouldSignalTL.get();
         return this.shouldSignal;
     }
 
