From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sun, 23 Feb 2025 01:04:29 +0300
Subject: [PATCH] Option to disable saving of snowball and firework


diff --git a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index c9f7e68abccbc6b894a66b675952c973e22af8e6..0c67e8e23945a94de6bd1af9188ff66c5d9bea29 100644
--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -350,4 +350,14 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
         double dz = hurtEntity.position().z - this.position().z;
         return DoubleDoubleImmutablePair.of(dx, dz);
     }
+
+    // DivineMC start - Option to disable saving firework
+    @Override
+    public boolean shouldBeSaved() {
+        if (this.level().divineConfig.disableFireworkSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // DivineMC end - Option to disable saving firework
 }
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index f8ca088ab75e6bfdfc566c97879c12beaeefcda0..8136f8dd6bf0fc9ad8f013804ee1a017e454cc2d 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -96,4 +96,14 @@ public class Snowball extends ThrowableItemProjectile {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
+
+    // DivineMC start - Option to disable snowball saving
+    @Override
+    public boolean shouldBeSaved() {
+        if (this.level().divineConfig.disableSnowballSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // DivineMC end - Option to disable snowball saving
 }
