From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sun, 23 Feb 2025 01:04:29 +0300
Subject: [PATCH] Option to disable saving of snowball and firework


diff --git a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index 325bdcd8c7dd4e03aa6f649c8f76b031529d5fd8..a7dd1efd8ab20e912d0da886a688db81471fa939 100644
--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -350,4 +350,14 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
         double dz = hurtEntity.position().z - this.position().z;
         return DoubleDoubleImmutablePair.of(dx, dz);
     }
+
+    // DivineMC start - Option to disable saving firework
+    @Override
+    public boolean shouldBeSaved() {
+        if (this.level().divineConfig.disableFireworkSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // DivineMC end - Option to disable saving firework
 }
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index 7735ba697e9f172ebedaae80fe4ad4cd48a8c558..209f5903d93f040804b56ebe692916908d353be9 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -97,4 +97,14 @@ public class Snowball extends ThrowableItemProjectile {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
+
+    // DivineMC start - Option to disable snowball saving
+    @Override
+    public boolean shouldBeSaved() {
+        if (this.level().divineConfig.disableSnowballSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // DivineMC end - Option to disable snowball saving
 }
