From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Mon, 7 Jul 2025 04:21:38 +0300
Subject: [PATCH] Small optimization to LinearPalette


diff --git a/net/minecraft/world/level/chunk/LinearPalette.java b/net/minecraft/world/level/chunk/LinearPalette.java
index 58f03cf8b32b141dc7df72568ef902eed53f4a5d..3c588d87410ff3615cf8d29310a9e73bdf8d6ffb 100644
--- a/net/minecraft/world/level/chunk/LinearPalette.java
+++ b/net/minecraft/world/level/chunk/LinearPalette.java
@@ -10,7 +10,7 @@ import org.apache.commons.lang3.Validate;
 public class LinearPalette<T> implements Palette<T>, ca.spottedleaf.moonrise.patches.fast_palette.FastPalette<T> { // Paper - optimise palette reads
     private final T[] values;
     private final int bits;
-    private int size;
+    private volatile int size; // DivineMC - Small optimization to LinearPalette
 
     // Paper start - optimise palette reads
     @Override
@@ -45,11 +45,16 @@ public class LinearPalette<T> implements Palette<T>, ca.spottedleaf.moonrise.pat
 
     @Override
     public int idFor(final T value, final PaletteResize<T> resizeHandler) {
-        for (int i = 0; i < this.size; i++) {
-            if (this.values[i] == value) {
+        // DivineMC start - Small optimization to LinearPalette
+        final T[] values = this.values;
+        final int currentSize = this.size;
+
+        for (int i = 0; i < currentSize; i++) {
+            if (values[i] == value) {
                 return i;
             }
         }
+        // DivineMC end - Small optimization to LinearPalette
 
         int index = this.size;
         if (index < this.values.length) {
@@ -63,17 +68,23 @@ public class LinearPalette<T> implements Palette<T>, ca.spottedleaf.moonrise.pat
 
     @Override
     public boolean maybeHas(final Predicate<T> predicate) {
-        for (int i = 0; i < this.size; i++) {
-            if (predicate.test(this.values[i])) {
+        // DivineMC start - Small optimization to LinearPalette
+        final T[] values = this.values;
+        final int currentSize = this.size;
+
+        for (int i = 0; i < currentSize; i++) {
+            T value = values[i];
+            if (value != null && predicate.test(value)) {
                 return true;
             }
         }
+        // DivineMC end - Small optimization to LinearPalette
 
         return false;
     }
 
     @Override
-    public T valueFor(final int index) {
+    public synchronized T valueFor(final int index) { // DivineMC - Small optimization to LinearPalette
         if (index >= 0 && index < this.size) {
             return this.values[index];
         } else {
