From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sat, 1 Feb 2025 18:38:26 +0300
Subject: [PATCH] Lag compensation


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index f5d4d48ad2b8b33a0be52bf935e371ea98d9183a..863ba3ffcd49cf28aa43b59b9b76bbdd05411699 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -338,6 +338,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public static final long SERVER_INIT = System.nanoTime(); // Paper - Lag compensation
     public boolean lagging = false; // Purpur - Lagging threshold
     protected boolean upnp = false; // Purpur - UPnP Port Forwarding
+    public final org.bxteam.divinemc.util.tps.TPSCalculator tpsCalculator = new org.bxteam.divinemc.util.tps.TPSCalculator(); // DivineMC - Lag compensation
     // Paper start - improve tick loop
     public final ca.spottedleaf.common.time.TickData tickTimes1s  = new ca.spottedleaf.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(1L));
     public final ca.spottedleaf.common.time.TickData tickTimes5s  = new ca.spottedleaf.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(5L));
@@ -1660,6 +1661,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         }
 
         this.server.spark.tickStart(); // Paper - spark
+        this.tpsCalculator.doTick(); // DivineMC - Lag compenstation
         new com.destroystokyo.paper.event.server.ServerTickStartEvent(this.tickCount+1).callEvent(); // Paper - Server Tick Events
         this.tickCount++;
         this.tickRateManager.tick();
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index cfdb8598054bc99bb9542d4d2c217c474e1c5e42..d3f3dd9aba7676b43191db45bd48b63ddafbc0c1 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -239,6 +239,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current)
     public boolean hasRidableMoveEvent = false; // Purpur - Ridables
     public net.minecraft.world.item.ItemStack ominousBanner; // DivineMC - Optimize Raids
+    public org.bxteam.divinemc.util.tps.TPSCalculator tpsCalculator = new org.bxteam.divinemc.util.tps.TPSCalculator(); // DivineMC - Lag Compensation
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
@@ -860,6 +861,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             }
         }
 
+        this.tpsCalculator.doTick(); // DivineMC - Lag compensation
+
         this.updateSkyBrightness();
         if (runs) {
             this.tickTime();
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 7fa156e3ddfc1d7896915bb6e1cc4a1961e0baca..ba8c4f42438c3ea75136081173642262629c818c 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -543,6 +543,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
         }
 
+        lagCompensation(); // DivineMC - Lag Compensation
         this.tickEffects();
         this.yHeadRotO = this.yHeadRot;
         this.yBodyRotO = this.yBodyRot;
@@ -551,6 +552,17 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         profiler.pop();
     }
 
+    // DivineMC start - Lag Compensation
+    private void lagCompensation() {
+        if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.lagCompensationEnabled || !org.bxteam.divinemc.config.DivineConfig.MiscCategory.potionEffectAcceleration) return;
+        if (this.level().isClientSide()) return;
+
+        for (int i = 0; i < ((ServerLevel) this.level()).tpsCalculator.applicableMissedTicks(); i++) {
+            tickEffects();
+        }
+    }
+    // DivineMC end - Lag Compensation
+
     protected boolean shouldTakeDrowningDamage() {
         return this.getAirSupply() <= -this.level().purpurConfig.drowningDamageInterval; // Purpur - Drowning Settings
     }
diff --git a/net/minecraft/world/entity/PortalProcessor.java b/net/minecraft/world/entity/PortalProcessor.java
index 7ec15f3d3a770e1e9e2fbac29dd38e947bb07091..a3c119c89aff9eb776b647a811ac9e4d8c0b7d25 100644
--- a/net/minecraft/world/entity/PortalProcessor.java
+++ b/net/minecraft/world/entity/PortalProcessor.java
@@ -24,10 +24,20 @@ public class PortalProcessor {
             return false;
         } else {
             this.insidePortalThisTick = false;
-            return allowedToTeleport && this.portalTime++ >= this.portal.getPortalTransitionTime(serverLevel, entity);
+            return allowedToTeleport && lagCompensation(this.portalTime++, serverLevel) >= this.portal.getPortalTransitionTime(serverLevel, entity); // DivineMC - Lag compensation
         }
     }
 
+    // DivineMC start - Lag compensation
+    private int lagCompensation(int original, ServerLevel world) {
+        if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.lagCompensationEnabled || !org.bxteam.divinemc.config.DivineConfig.MiscCategory.portalAcceleration) return original;
+        if (world.isClientSide()) return original;
+
+        portalTime = portalTime + world.tpsCalculator.applicableMissedTicks();
+        return portalTime;
+    }
+    // DivineMC end - Lag compensation
+
     public @Nullable TeleportTransition getPortalDestination(final ServerLevel serverLevel, final Entity entity) {
         return this.portal.getPortalDestination(serverLevel, entity, this.entryPosition);
     }
diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index cc6b346ca4beaae8d40a2b97e74c2cabafbfc8ad..dab7a08125afcc1c4e38d7668463f757eb7fa39a 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -154,8 +154,25 @@ public class ItemEntity extends Entity implements TraceableEntity {
     }
     // Paper end - EAR 2
 
+    // DivineMC start - Lag compensation
+    private void lagCompensation() {
+        if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.lagCompensationEnabled || !org.bxteam.divinemc.config.DivineConfig.MiscCategory.pickupAcceleration) return;
+        if ((this).level().isClientSide()) return;
+
+        if (pickupDelay == 0) return;
+
+        if (pickupDelay - ((ServerLevel) this.level()).tpsCalculator.applicableMissedTicks() <= 0) {
+            pickupDelay = 0;
+            return;
+        }
+
+        pickupDelay = pickupDelay - ((ServerLevel) this.level()).tpsCalculator.applicableMissedTicks();
+    }
+    // DivineMC end - Lag compensation
+
     @Override
     public void tick() {
+        lagCompensation(); // DivineMC - Lag compensation
         if (this.getItem().isEmpty()) {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
diff --git a/net/minecraft/world/item/Item.java b/net/minecraft/world/item/Item.java
index a6a5800f8c13977cedafc0cce95aed3a05306230..24d0c3fa857deb7fc6ea0bc9b89eef5a3723a620 100644
--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -307,10 +307,25 @@ public class Item implements ItemLike, FeatureElement {
         }
     }
 
+    // DivineMC start - Lag compensation
+    private int lagCompensation(int original, net.minecraft.server.level.ServerLevel level) {
+        if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.lagCompensationEnabled || !org.bxteam.divinemc.config.DivineConfig.MiscCategory.eatingAcceleration || original == 0) return original;
+        return org.bxteam.divinemc.util.tps.TPSUtil.tt20(original, true, level);
+    }
+    // DivineMC end - Lag compensation
+
     public int getUseDuration(final ItemStack itemStack, final LivingEntity user) {
         Consumable consumable = itemStack.get(DataComponents.CONSUMABLE);
         if (consumable != null) {
-            return consumable.consumeTicks();
+            // DivineMC start - Lag compensation
+            int original = consumable.consumeTicks();
+
+            if (user.level() instanceof net.minecraft.server.level.ServerLevel serverLevel) {
+                return lagCompensation(original, serverLevel);
+            }
+
+            return original;
+            // DivineMC end - Lag compensation
         } else {
             return !itemStack.has(DataComponents.BLOCKS_ATTACKS) && !itemStack.has(DataComponents.KINETIC_WEAPON) ? 0 : 72000;
         }
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 720e6c9669ae0ca9fed7ff62de20f18203e2338b..22cc900443644a65be570319779c536f6bee6e46 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -356,14 +356,22 @@ public abstract class BlockBehaviour implements FeatureElement {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
     }
 
+    // DivineMC start - Lag compensation
+    private float lagCompensation(float original, Player player) {
+        if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.lagCompensationEnabled || !org.bxteam.divinemc.config.DivineConfig.MiscCategory.blockBreakingAcceleration) return original;
+        if (player.level().isClientSide) return original;
+        return original * org.bxteam.divinemc.util.tps.TPSCalculator.MAX_TPS / (float) ((ServerLevel) player.level()).tpsCalculator.getMostAccurateTPS();
+    }
+    // DivineMC end - Lag compensation
+
     protected float getDestroyProgress(final BlockState state, final Player player, final BlockGetter level, final BlockPos pos) {
         float destroySpeed = state.getDestroySpeed(level, pos);
         if (destroySpeed == -1.0F) {
-            return 0.0F;
+            return lagCompensation(0.0F, player); // DivineMC - Lag compensation
         }
 
         int modifier = player.hasCorrectToolForDrops(state) ? 30 : 100;
-        return player.getDestroySpeed(state) / destroySpeed / modifier;
+        return lagCompensation(player.getDestroySpeed(state) / destroySpeed / modifier, player); // DivineMC - Lag compensation
     }
 
     protected void spawnAfterBreak(final BlockState state, final ServerLevel level, final BlockPos pos, final ItemStack tool, final boolean dropExperience) {
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 4c0e247cb9a453c0554de68c0fa803bb94108895..4625cb804bec4330b2c228ed5b65db4b20891455 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -945,6 +945,19 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
             this.ticker = ticker;
         }
 
+        // DivineMC start - Lag compensation
+        private <T extends BlockEntity> void lagCompensation(Runnable original) {
+            original.run();
+            if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.lagCompensationEnabled) return;
+            if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.blockEntityAcceleration) return;
+            if (LevelChunk.this.level.isClientSide()) return;
+
+            for (int i = 0; i < ((ServerLevel) this.blockEntity.getLevel()).tpsCalculator.applicableMissedTicks(); i++) {
+                original.run();
+            }
+        }
+        // DivineMC end - Lag compensation
+
         @Override
         public void tick() {
             if (!this.blockEntity.isRemoved() && this.blockEntity.hasLevel()) {
@@ -955,7 +968,11 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
                         profiler.push(this::getType);
                         BlockState blockState = LevelChunk.this.getBlockState(pos);
                         if (this.blockEntity.getType().isValid(blockState)) {
-                            this.ticker.tick(LevelChunk.this.level, this.blockEntity.getBlockPos(), blockState, this.blockEntity);
+                            // DivineMC start - Lag compensation
+                            lagCompensation(() -> {
+                                this.ticker.tick(LevelChunk.this.level, this.blockEntity.getBlockPos(), blockState, this.blockEntity);
+                            });
+                            // DivineMC end - Lag compensation
                             this.loggedInvalidBlockState = false;
                             // Paper start - Remove the Block Entity if it's invalid
                         } else {
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index f5baf1490f98a0469967090b06e39bdd0855f3ed..d67d81cfb9388bbb915831ffc7409ef64d2847b2 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -189,9 +189,22 @@ public abstract class LavaFluid extends FlowingFluid {
         return state.getHeight(level, pos) >= 0.44444445F && other.is(FluidTags.WATER);
     }
 
+    // DivineMC start - Lag compensation
+    private int lagCompensation(int original, ServerLevel level) {
+        if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.lagCompensationEnabled || !org.bxteam.divinemc.config.DivineConfig.MiscCategory.fluidAcceleration) return original;
+        return org.bxteam.divinemc.util.tps.TPSUtil.tt20(original, true, level);
+    }
+    // DivineMC end - Lag compensation
+
     @Override
     public int getTickDelay(final LevelReader level) {
-        return level.getWorldBorder().world != null ? (isFastLava(level) ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether) : (isFastLava(level) ? 10 : 30); // Purpur - Make lava flow speed configurable
+        // DivineMC start - Lag compensation
+        int original = level.getWorldBorder().world != null ? (isFastLava(level) ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether) : (isFastLava(level) ? 10 : 30); // Purpur - Make lava flow speed configurable
+        if (level instanceof ServerLevel serverLevel) {
+            return lagCompensation(original, serverLevel);
+        }
+        return original;
+        // DivineMC end - Lag compensation
     }
 
     @Override
diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index e52362ae20646f0311262853af0cdc690a603b52..15c6f3d9575344595f5e8513e1e30920a0e24df9 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -123,8 +123,16 @@ public abstract class WaterFluid extends FlowingFluid {
         return 1;
     }
 
+    // DivineMC start - Lag compensation
+    private int lagCompensation(ServerLevel level) {
+        if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.lagCompensationEnabled || !org.bxteam.divinemc.config.DivineConfig.MiscCategory.fluidAcceleration) return 5;
+        return org.bxteam.divinemc.util.tps.TPSUtil.tt20(5, true, level);
+    }
+    // DivineMC end - Lag compensation
+
     @Override
     public int getTickDelay(final LevelReader level) {
+        if (level instanceof ServerLevel serverLevel) return lagCompensation(serverLevel); // DivineMC - Lag compensation
         return 5;
     }
 
