From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sat, 1 Feb 2025 18:55:59 +0300
Subject: [PATCH] MSPT Tracking for each world


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 6b51ed6c74c711d26e077ae75b88ea180170efa7..309cb08f40e7a07367d3fe3727beccb8d57102d1 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1826,7 +1826,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // DivineMC start - Parallel world ticking
     private void tickLevel(ServerLevel serverLevel, BooleanSupplier haveTime) {
         try {
+            // DivineMC start - MSPT Tracking for each world
+            long tickStart = Util.getNanos();
             serverLevel.tick(haveTime);
+
+            long tickDuration = Util.getNanos() - tickStart;
+
+            serverLevel.tickTimes5s.add(this.tickCount, tickDuration);
+            serverLevel.tickTimes10s.add(this.tickCount, tickDuration);
+            serverLevel.tickTimes60s.add(this.tickCount, tickDuration);
+            // DivineMC end - MSPT Tracking for each world
         } catch (Throwable levelTickingException) {
             CrashReport crashReport = CrashReport.forThrowable(levelTickingException, "Exception ticking world");
             serverLevel.fillReportDetails(crashReport);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 951298d3a355eca1b31e93dc050d0bf8344da126..2ba7322e996d12e0ed77966af95879177c2cbe49 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -611,6 +611,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
     // Paper end - chunk tick iteration
 
+    // DivineMC start - MSPT Tracking for each world
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+    // DivineMC end - MSPT Tracking for each world
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
