From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Thu, 10 Jul 2025 04:30:21 +0300
Subject: [PATCH] Copper Bulb 1gt delay


diff --git a/net/minecraft/world/level/block/CopperBulbBlock.java b/net/minecraft/world/level/block/CopperBulbBlock.java
index ebc44c7adf15ea8a67b4253a18eafe2c9eadb3e2..ff959de46dc39fca54e5ef25bb8caef5a9502ccc 100644
--- a/net/minecraft/world/level/block/CopperBulbBlock.java
+++ b/net/minecraft/world/level/block/CopperBulbBlock.java
@@ -33,7 +33,13 @@ public class CopperBulbBlock extends Block {
     @Override
     protected void onPlace(final BlockState state, final Level level, final BlockPos pos, final BlockState oldState, final boolean movedByPiston) {
         if (oldState.getBlock() != state.getBlock() && level instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // DivineMC start - Copper Bulb 1gt delay
+            if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                level.scheduleTick(pos, this, 1);
+            }
+            // DivineMC end - Copper Bulb 1gt delay
         }
     }
 
@@ -42,9 +48,23 @@ public class CopperBulbBlock extends Block {
         final BlockState state, final Level level, final BlockPos pos, final Block block, final @Nullable Orientation orientation, final boolean movedByPiston
     ) {
         if (level instanceof ServerLevel serverLevel) {
-            this.checkAndFlip(state, serverLevel, pos);
+            // DivineMC start - Copper Bulb 1gt delay
+            if (!org.bxteam.divinemc.config.DivineConfig.MiscCategory.copperBulb1gt) {
+                this.checkAndFlip(state, serverLevel, pos);
+            } else {
+                level.scheduleTick(pos, this, 1);
+            }
+        }
+    }
+
+    // DivineMC start - Copper Bulb 1gt delay
+    @Override
+    public void tick(BlockState state, ServerLevel level, BlockPos pos, net.minecraft.util.RandomSource random) {
+        if (org.bxteam.divinemc.config.DivineConfig.MiscCategory.copperBulb1gt) {
+            checkAndFlip(state, level, pos);
         }
     }
+    // DivineMC end - Copper Bulb 1gt delay
 
     public void checkAndFlip(final BlockState state, final ServerLevel level, final BlockPos pos) {
         boolean signal = level.hasNeighborSignal(pos);
