From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Fri, 11 Jul 2025 20:26:35 +0300
Subject: [PATCH] Update Purpur rebrand settings


diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index a337df7cdb4779fea9bbab039dcfab4aad6fa939..7b84a68cb7656b72eb23260679ecae16333b0a66 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -230,10 +230,16 @@ public class PurpurConfig {
         deathMessageOnlyBroadcastToAffectedPlayer = getBoolean("settings.broadcasts.death.only-broadcast-to-affected-player", deathMessageOnlyBroadcastToAffectedPlayer);
     }
 
+    // DivineMC start - Update Purpur rebrand settings
     public static String serverModName = io.papermc.paper.ServerBuildInfo.buildInfo().brandName();
-    private static void serverModName() {
-        serverModName = getString("settings.server-mod-name", serverModName);
-    }
+    public static String serverGUIName = io.papermc.paper.ServerBuildInfo.buildInfo().brandName() + " Console";
+    public static String f3Name = io.papermc.paper.ServerBuildInfo.buildInfo().brandName();
+    private static void serverName() {
+        serverModName = getString("settings.rebrand.server-mod-name", serverModName);
+        serverGUIName = getString("settings.rebrand.server-gui-name", serverGUIName);
+        f3Name = getString("settings.rebrand.f3-name", f3Name);
+    }
+    // DivineMC end - Update Purpur rebrand settings
 
     public static double laggingThreshold = 19.0D;
     private static void tickLoopSettings() {
