--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -974,7 +_,7 @@
 
     @Override
     public List<World> getWorlds() {
-        return new ArrayList<World>(this.worlds.values());
+        return new it.unimi.dsi.fastutil.objects.ObjectArrayList<World>(this.worlds.values()); // DivineMC - optimize getWorlds
     }
 
     @Override
